### PR TITLE
Make sure input numpy array is aligned.

### PR DESCRIPTION
**Pull request by trivialfis**
- use `np.require` to specify that the alignment is required.
- scipy csr as well.
- validate input pointer in `ArrayInterface`.

### CODE REVIEW

File: src/data/array_interface.h [hunk 1]
```diff
@@ -400,6 +401,13 @@ class ArrayInterface {
 
     data = ArrayInterfaceHandler::ExtractData(array, n);
     static_assert(allow_mask ? D == 1 : D >= 1, "Masked ndarray is not supported.");
+
+    this->DispatchCall([&](auto const *values) {
```
Comment (hcho3) [inline src/data/array_interface.h:405]:
```suggestion
    this->DispatchCall([&](auto const *data_casted) {
```
to indicate that the parameter is really a casted version of `data`.

File: tests/python/test_dmatrix.py [hunk 2]
```diff
@@ -327,7 +327,7 @@ def test_sparse_dmatrix_csr(self):
         nrow = 100
         ncol = 1000
         x = rand(nrow, ncol, density=0.0005, format='csr', random_state=rng)
-        assert x.indices.max() < ncol - 1
+        assert x.indices.max() <= ncol - 1
```
Comment (hcho3) [inline tests/python/test_dmatrix.py:330]:
```suggestion
        assert x.indices.max() < ncol
```
I think this is slightly more idiomatic.

**Review by hcho3 — CHANGES_REQUESTED** [review]

File: src/data/array_interface.h (hunk 1 above)
Comment (trivialfis) [reply src/data/array_interface.h:405]: @hcho3 Not casted version is `void const*`, which is actually casted from input pointer (user-input-ptr -> void const* -> user-input-ptr)

Comment (hcho3) [reply src/data/array_interface.h:405]: How about `data_typed_ptr`?

Comment (trivialfis) [reply src/data/array_interface.h:405]: sounds good.

Comment (trivialfis) [reply src/data/array_interface.h]: done.

File: tests/python/test_dmatrix.py (hunk 2 above)
Comment (trivialfis) [reply tests/python/test_dmatrix.py]: done.

**Review by hcho3 — APPROVED** [review]

**Pull request merged**